include/RAJA/policy/cuda/nested.hpp

```diff
@@ -113,30 +113,150 @@ struct Executor<ForTypeIn<Index, cuda_exec<block_size>, Rest...>> {
 };
 
 
-struct cuda_collapse_exec{};
+namespace internal {
 
-template <typename FT0, typename FT1>
-struct Executor<Collapse<cuda_collapse_exec, FT0, FT1>> {
 
-  // TODO, check that FT0 and FT1 are cuda policies
+
+/*!
+ * \brief Launcher that uses execution policies to map blockIdx and threadIdx to
+ * map
+ * to N-argument function
+ */
+template <typename BODY, typename... CARGS>
+__global__ void cudaLauncher(BODY loop_body, CARGS... cargs)
+{
+  // force reduction object copy constructors and destructors to run
+  auto body = loop_body;
+
+  // Invoke the wrapped body.
+  // The wrapper will take care of computing indices, and deciding if the
+  // given block+thread is in-bounds, and invoking the users loop body
+  body();
+}
+
+
+} // namespace internal
+
+
+template<bool Async = false>
+struct cuda_collapse_exec : public cuda_exec<0, Async>
+{};
+
+
+template<typename ... FOR>
+using CudaCollapse = Collapse<cuda_collapse_exec<false>, FOR ...>;
+
+template<typename ... FOR>
+using CudaCollapseAsync = Collapse<cuda_collapse_exec<true>, FOR ...>;
+
+
+// TODO, check that FT... are cuda policies
+template <bool Async, typename ... FT>
+struct Executor<Collapse<cuda_collapse_exec<Async>, FT ...>> {
+
+  using collapse_policy = Collapse<cuda_collapse_exec<Async>, FT...>;
+
+
+
+  template <typename BaseWrapper, typename ... LoopPol>
+  struct ForWrapper {
+    using data_type = typename BaseWrapper::data_type;
+    data_type data;
+    camp::tuple<LoopPol...> loop_policies;
+    using ft_tuple = camp::list<FT...>;
+    using index_sequence = typename camp::make_idx_seq<sizeof...(LoopPol)>::type;
+
+    // Explicitly unwrap the data from the wrapper
+    ForWrapper(BaseWrapper const &w, LoopPol &&... pol) :
+      data(w.data),
+      loop_policies(camp::make_tuple(std::forward<LoopPol>(pol)...))
+    {}
+
+
+    /*
+     * Evaluates the loop index for the idx'th loop in the Collapse
+     */
+    template<size_t idx>
+    RAJA_DEVICE
+    RAJA::Index_type evalLoopIndex(){
+      // grab the loop policy
+      auto &policy = camp::get<idx>(loop_policies);
+
+      // grab the For type from our type list
+      using ft = typename camp::at_v<ft_tuple, idx>::type;
+
+      // Assign the For index value to the correct argument
+      int loop_value = policy();
+      data.template assign_index<ft::index_val>( loop_value );
```

Comment:
Ugh... that really needs a better interface doesn't it? I'll think about how to make this a little less cumbersome.